web/js/Model.js

```diff
@@ -0,0 +1,200 @@
+        this.idx_offset = 0;//how many has poped out
+        
+        this.scatterLayout = [];
+        this.stat = [];
+
+        this.func_names = []
+        this.prog_names = []
+        this.initial_timestamp = -1
+        this.prev_receive_time = -1
+        this.global_rank_anomaly = {}
+
+        this.delta = {};
+        this.prev = {};
+        this.frameID = 0;
+        this.frameWindow = 30
+        
+        this.frames = {};
+        
+        this.setWait = true;
+        this.NUM_SELECTION_RANK = 10;
+        this.history = {};
+        this.selectedExecution = {};
+        this.selectedRankInfo = {};
+
+    }
+    update(stream) {
+        /**
+         * concat new array to the corresponding array based on the rank
+         */
+        for(var rank in stream) { 
+            if(!this.frames[rank]) {
+                this.frames[rank] = []
+            }
+            this.frames[rank] = this.frames[rank].concat(stream[rank]) 
+        }
+    }
+    getSortedRanks(delta) {
+        /**
+         * Sorts delta values per rank 
+         * Get top and bottom 5 ranks based on delta values
+         */
+        var ranks = Object.keys(delta)
+        var deltaValues = Object.values(delta);
+        var sortedRanks = deltaValues.map((d, i) => [ranks[i], d]) 
+                        .sort(([r1, d1], [r2, d2]) => d2 - d1) 
+                        .map(([r, d]) => r); 
+        var top;
+        var bottom;
+        if( sortedRanks.length < (this.NUM_SELECTION_RANK*2) ) {
+            var m = Math.floor((sortedRanks.length)/2) // adjust the number of ranks is under 10
+            top = sortedRanks.slice(0, m)
+            bottom = sortedRanks.slice(m)
+        } else {
+            top = sortedRanks.slice(0, this.NUM_SELECTION_RANK)
+            bottom = sortedRanks.slice(sortedRanks.length-this.NUM_SELECTION_RANK)
+        }
+        return {
+            'top': top,
+            'bottom': bottom
+        }
+    }
+    hasReceived() {
+        /**
+         * Return true if possible to process
+         * If possible, 
+         *      Gets the first element, the number of anomalies, from each list of rank
+         *      Considers only top and bottom sorted ranks
+         */
+        // var curr_delta = this.updateDelta();
+        this.updateDelta();
+        this.selectedRanks = this.getSortedRanks(this.delta)
+        var res = false;
+        for ( var rank in this.frames) {
+            var rankData = this.frames[rank]
+            if (rankData.length > 0) {
+                var value = rankData.splice(0, 1)[0]
+                if(this.selectedRanks.top.includes(rank) || this.selectedRanks.bottom.includes(rank)) {
+                    res = true;
+                }
+                if (!this.history[this.frameID]) {
+                    this.history[this.frameID] = {} 
+                }
+                this.history[this.frameID][rank] = value;
+            }
+        }
+        return res;
+    }
+    updateDelta() {
```

Comment:
Using the scalable visualization server, updateDelta() will not work because 
1. statistics will be updated asynchronously and
2. it cannot guarantee the current frame is the frame right after the frame the front-end code has.
3. so that this part should be re-considered or need to consider to use standard deviation rather that delta.

Reply:
updateDelta() is temporal (to assume the backend) and should be removed after integration. AD and PS will be requested a type of statistic and comparison size.